WMF Framework/NavigationState.swift

```diff
@@ -0,0 +1,95 @@
+public struct NavigationState: Codable {
+    static let libraryKey = "nav_state"
+
+    public var viewControllers: [ViewController]
+    public var shouldAttemptLogin: Bool
+
+    public init(viewControllers: [ViewController], shouldAttemptLogin: Bool) {
+        self.viewControllers = viewControllers
+        self.shouldAttemptLogin = shouldAttemptLogin
+    }
+    
+    public struct ViewController: Codable {
+        public var kind: Kind
+        public var presentation: Presentation
+        public var info: Info?
+        public var children: [ViewController]
+        
+        public mutating func updateChildren(_ children: [ViewController]) {
+            self.children = children
+        }
+        
+        public init?(kind: Kind?, presentation: Presentation, info: Info? = nil, children: [ViewController] = []) {
+            guard let kind = kind else {
+                return nil
+            }
+            self.kind = kind
+            self.presentation = presentation
+            self.info = info
+            self.children = children
+        }
+        
+        public enum Kind: Int, Codable {
```

Comment:
> Is there a better way to structure saving/restoring nav state?

One thought is you could try splitting up the info struct into kind-specific info structs, then pass them in as associated values to this Kind enum. That way you don't have to deal with all the optionals and can guarantee it's set up with all the info it needs to restore itself if the correct kind is set. Not sure if that will give you grief with Codable though.

Reply:
associated values won't work with Codable 😞 I saw workarounds (https://www.objc.io/blog/2018/01/23/codable-enums/) but I thought it wasn't worth the manual encoding